sycl/include/sycl/queue.hpp

```diff
@@ -1059,6 +1059,119 @@ class __SYCL_EXPORT queue {
         CodeLoc);
   }
 
+  /// Copies data from a memory region pointed to by a placeholder accessor to
+  /// another memory region pointed to by a shared_ptr.
+  ///
+  /// \param Src is a placeholder accessor to the source memory.
+  /// \param Dest is a shared_ptr to the destination memory.
+  /// \return an event representing copy operation.
+  template <typename SrcT, int SrcDims, access_mode SrcMode, target SrcTgt,
+            access::placeholder IsPlaceholder, typename DestT>
+  event copy(accessor<SrcT, SrcDims, SrcMode, SrcTgt, IsPlaceholder> Src,
+             std::shared_ptr<DestT> Dest) {
```

Comment:
Since this is a top-level API that calls submit, it would be nice to capture and pass the code-location with it. See for example `ext_oneapi_submit_barrier`.

Arguably we should do the same for some of the other existing memory operations, but that can be in another patch.

Reply:
No problem, I'll add that to these functions

Reply:
I imagine this would just be adding `_CODELOCPARAM(CodeLoc)` to the param list and passing it to the `submit` call via `_CODELOCFW(CodeLoc)`?

Wouldn't this change the parameter list and be different to the spec? (I suppose getting these to call private helper functions with the code location would be the way to go?)

Reply:
> I imagine this would just be adding `_CODELOCPARAM(CodeLoc)` to the param list and passing it to the `submit` call via `_CODELOCFW(CodeLoc)`?

Exactly. We use these for instrumentation.

> Wouldn't this change the parameter list and be different to the spec? (I suppose getting these to call private helper functions with the code location would be the way to go?)

Technically they do change the parameters of the functions, but they need to be done at top-level to correctly capture the origin location of the calls. Note that they have default arguments and can be disabled with `DISABLE_SYCL_INSTRUMENTATION_METADATA` so in general it is okay to add.